### PR TITLE
feat: C509 `AlgorithmIdentifier` for `subjectPublicKeyAlgorithm` and `issuerSignatureAlgorithm`

### PR DESCRIPTION
**Pull request by bkioshn**
# Description

Implement encode and decode for `AlgorithmIdentifier` for `subjectPublicKeyAlgorithm` and `issuerSignatureAlgorithm`

## Related Issue(s)
#606 

## Description of Changes

- Implement encode and decode for `AlgorithmIdentifier` where
```
 AlgorithmIdentifier = int / ~oid /
                        [ algorithm: ~oid, parameters: bytes ]
```
 Since `subjectPublicKeyAlgorithm` and `issuerSignatureAlgorithm` use different registry table, 
they are both handle separately for `int` value encoding and decoding
- Add test cases

## Please confirm the following checks

* [x] My code follows the style guidelines of this project
* [x] I have performed a self-review of my code
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation
* [x] My changes generate no new warnings
* [x] I have added tests that prove my fix is effective or that my feature works
* [x] New and existing unit tests pass locally with my changes
* [x] Any dependent changes have been merged and published in downstream module

### CODE REVIEW

**Review by stevenj — APPROVED** [review]
Looks Great

File: catalyst-gateway-crates/c509-certificate/src/c509_issuer_sig_algo/data.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    (29,    oid!(1.3.6.1.5.5.7.6.30),            "RSASSA-PSS with SHAKE128"),
+    (30,    oid!(1.3.6.1.5.5.7.6.3),             "RSASSA-PSS with SHAKE256"),
+    (42,    oid!(1.2.840.113549.1.9.16.3.17),    "HSS / LMS"),
+    (43,    oid!(0.4.0.127.0.15.1.1.13.0),       "XMSS"),
+    (44,    oid!(0.4.0.127.0.15.1.1.14.0),       "XMSS^MT"),
+    (45,    oid!(1.2.156.10197.1.501),           "SM2 with SM3"),    
+];
+
+/// A struct of data that contains lookup tables for `IssuerSignatureAlgorithm`.
+pub(crate) struct IssuerSigAlgoData {
```
Comment (Mr-Leshiy) [inline catalyst-gateway-crates/c509-certificate/src/c509_issuer_sig_algo/data.rs:48]: Better to change it to 
```Rust
pub(crate) struct IssuerSigAlgoData(IntegerToOidTable)
```

File: catalyst-gateway-crates/c509-certificate/src/c509_issuer_sig_algo/data.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    for data in SIG_ALGO_DATA {
+        int_to_oid_table.add(data.0, data.1);
+    }
+
+    IssuerSigAlgoData { int_to_oid_table }
+});
+
+/// Static reference to the `IssuerSigAlgoData` lookup table.
+pub(crate) static ISSUER_SIG_ALGO_LOOKUP: &Lazy<IssuerSigAlgoData> = &ISSUER_SIG_ALGO_TABLE;
```
Comment (Mr-Leshiy) [inline catalyst-gateway-crates/c509-certificate/src/c509_issuer_sig_algo/data.rs:72]: why we need this variable as it is just a reference of the `ISSUER_SIG_ALGO_TABLE` ?
Seems redundant to me, could we remove it and just use a `ISSUER_SIG_ALGO_TABLE` ?

File: catalyst-gateway-crates/c509-certificate/src/c509_subject_pub_key_algo/data.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    for data in PUB_KEY_ALGO_DATA {
+        int_to_oid_table.add(data.0, data.1);
+    }
+
+    SubjectPubKeyAlgoData { int_to_oid_table }
+});
+
+/// Static reference to the `SubjectPubKeyAlgoData` lookup table.
+pub(crate) static SUBJECT_PUB_KEY_ALGO_LOOKUP: &Lazy<SubjectPubKeyAlgoData> =
```
Comment (Mr-Leshiy) [inline catalyst-gateway-crates/c509-certificate/src/c509_subject_pub_key_algo/data.rs:64]: same

**Review by Mr-Leshiy — CHANGES_REQUESTED** [review]

File: catalyst-gateway-crates/c509-certificate/src/c509_issuer_sig_algo/data.rs (hunk 2 above)
Comment (bkioshn) [reply catalyst-gateway-crates/c509-certificate/src/c509_issuer_sig_algo/data.rs:72]: the `Lazy` static should be a private function, 
```
static ISSUER_SIG_ALGO_TABLE: Lazy<IssuerSigAlgoData> = Lazy::new(|| {
```
but somehow I found out that it need to be called to get the table inside, don't want to end up doing something like 
```
get_extension_data().get_int_to_oid_table()
```
I will come back and refactor it later
@Mr-Leshiy

File: catalyst-gateway-crates/c509-certificate/src/c509_algo_identifier.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+    /// Get the OID.
+    pub(crate) fn get_oid(&self) -> Oid<'static> {
+        self.oid.clone().get_oid()
+    }
+
+    /// Get the parameter.
+    pub(crate) fn get_param(&self) -> Option<String> {
+        self.param.clone()
+    }
```
Comment (apskhem) [inline catalyst-gateway-crates/c509-certificate/src/c509_algo_identifier.rs:44]: IMO, these getter methods shouldn't handle the clone by themselves. It'd be better if cloning is decided by the caller.

File: catalyst-gateway-crates/c509-certificate/src/c509_algo_identifier.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    /// Get the OID.
+    pub(crate) fn get_oid(self) -> Oid<'static> {
+        self.oid.get_oid()
+    }
+
+    /// Get the parameter.
+    pub(crate) fn get_param(self) -> Option<String> {
+        self.param
+    }
```
Comment (apskhem) [inline catalyst-gateway-crates/c509-certificate/src/c509_algo_identifier.rs:44]: Shouldn't comsume itself just to read a value. You can make it returns a reference of a value with `&` instead.

**Review by apskhem — APPROVED** [review]

**Review by Mr-Leshiy — APPROVED** [review]

**Pull request merged**